avm/src/vm/memory.rs

```diff
@@ -876,15 +878,17 @@ impl HandlerMemory {
   /// Takes a fractal memory and treats it like a UTF-8 encoded Pascal string, and the converts it
   /// to something Rust can work with. This function *may* crash if the underlying data is not a
   /// UTF-8 encoded Pascal string.
-  pub fn fractal_to_string(f: FractalMemory) -> String {
+  pub fn fractal_to_string(f: FractalMemory) -> VMResult<String> {
     let s_len = f.block[0].1 as usize;
-    let mut s_bytes: Vec<u8> = Vec::new();
-    for i in 1..f.block.len() {
-      let mut b = f.block[i].1.to_ne_bytes().to_vec();
-      s_bytes.append(&mut b);
-    }
-    let s = str::from_utf8(&s_bytes[0..s_len]).unwrap();
-    s.to_string()
+    let s_bytes = f
+      .block
+      .iter()
+      .skip(1)
+      // TODO: `IntoIter::new` will be deprecated once `rust-lang/rust#65819` is merged
+      .flat_map(|(_, chars)| IntoIter::new(chars.to_ne_bytes()))
+      .collect::<Vec<u8>>();
```

Comment:
Why switch from the `for in` loop to this if it depends on discourage usage of a built-in type?

Reply:
the `for in` loop from before had a non-constant number of allocations - this iterator implementation has 1 allocation total, guaranteed (on `iter()` - `skip` doesn't allocate, `IntoIter::new` recasts the array from `chars.to_ne_bytes` to an iterator, and `collect` inherits the allocation from `iter()`)

Reply:
(side note: `impl IntoIterator for [T; N]` will be stabilized soon, but it'll actually call `impl IntoIter for Slice`, which will result in dropping the array while returning the iterator on a slice of the array, which is a compiler error. In edition 2021, this will delegate to `array::IntoIter`, which won't be deprecated)